### PR TITLE
reef: mon/ElectionLogic: tie-breaker mon ignore proposal from marked down mon

### PR DESCRIPTION
**Pull request by kamoltat**
### To view full analysis here:

https://docs.google.com/document/d/1YW7RKTUlMw2nxB9vzg_7Dp32i0Tf7i1WOt6gvyZOV-g/edit?usp=sharing

### Problem:

Monitor election gets stuck,
resulting in the cluster being
unaccessible.

Set up stretch cluster:
DC1: mon.a,mon.b
DC2: mon.c, mon.d
Arbiter: mon.e

Apply netsplit between
DC1 and DC2, wait around
10 seconds, client tried accessing
the cluster ... failed to access.

We expect the cluster to be functional
when there is netsplit.

This was due to how we suppose to
kick one DC out of quorum and keep them in the
dark until the connection comes back (netsplit gone).
However, there is a small window where the out-of-quorum
DC is able to have an influence on the tiebreaker MON
by forcing the tiebreaker MON to bump its epoch
which then it will reject the winner MON's victory
message from in-quorum DC.

### Solution:
In CONNECTIVITY election strategy,
tie-breaker MON disregard any proposal that is coming
from MONs that belongs to `stretch_marked_down_mons`
set. As a result, the out-of-quorum
MONs won't get the chance to force the tie-breaker
MON to increase its epoch, hence will now accept
the victory message from the in-quorum Monitors.

Fixes: https://tracker.ceph.com/issues/66471

Backport Tracker: https://tracker.ceph.com/issues/67050

Backporting relevant commits from main PR(s):
https://github.com/ceph/ceph/pull/57906

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by kamoltat** [comment]
jenkins test make check

File: src/test/mon/test_election.cc [hunk 1]
```diff
@@ -187,6 +189,18 @@ struct Owner : public ElectionOwner, RankProvider {
     quorum = members;
     victory_accepters = 1;
   }
+  bool is_stretch_marked_down_mons(int rank) const {
```
Comment (athanatos) [inline src/test/mon/test_election.cc:192]: return stretch_marked_down_mons.contains(rank);

**Review by athanatos — APPROVED** [review]
LGTM (ignore my previous comment, didn't notice this was a backport)

**Comment by kamoltat** [comment]
Scheduled against RADOS suite 277 jobs in total:
https://pulpito.ceph.com/ksirivad-2024-08-30_14:53:30-rados-wip-ksirivad-backport-reef-57906-distro-default-smithi/

will review once the jobs finished.

**Comment by kamoltat** [comment]
RADOS APPROVED:
https://pulpito.ceph.com/ksirivad-2024-08-30_14:53:30-rados-wip-ksirivad-backport-reef-57906-distro-default-smithi/

Failures, Unrelated
- https://tracker.ceph.com/issues/65183 -  Overriding an EC pool needs the "--yes-i-really-mean-it" flag in addition to "force"
- https://tracker.ceph.com/issues/59196 - ceph_test_lazy_omap_stats segfault while waiting for active+clean
- https://tracker.ceph.com/issues/62776 - rados: cluster [WRN] overall HEALTH_WARN - do not have an application enabled
- https://tracker.ceph.com/issues/67684 - cephadm: orch/cephadm suite tasks using nvme-loop on centos 9 failing on `nvme connect` command
- https://tracker.ceph.com/issues/61774 - centos 9 testing reveals rocksdb "Leak_StillReachable" memory leak in mons
- https://tracker.ceph.com/issues/64208 - test_cephadm.sh: Container version mismatch causes job to fail.
- https://tracker.ceph.com/issues/47838 - mon/test_mon_osdmap_prune.sh: first_pinned != trim_to

**Pull request merged**